[Navigation] Dispatch NavigateEvent on same-document navigations

TingPing opened the pull request:
#### b3b7f859d975be6b081664c92fc245c650830cef
<pre>
[Navigation] Dispatch NavigateEvent on same-document navigations
<a href="https://bugs.webkit.org/show_bug.cgi?id=265401">https://bugs.webkit.org/show_bug.cgi?id=265401</a>

Reviewed by Alex Christensen.

- Partial implementation intercept/finish/scroll/abort in NavigateEvent.
- Implement a large part of the dispatching of Traversal and Push/Reload/Replace events.
- The FrameLoader dispatches the events in the case of same-document loads.
- Gate NavigateEvent.hasUAVisualTransition behind the setting for it.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/commit-behavior/multiple-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/click-samedocument-crossorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/click-samedocument-crossorigin-sameorigindomain.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/click-samedocument-sameorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/location-samedocument-crossorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/location-samedocument-crossorigin-sameorigindomain.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/location-samedocument-sameorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/open-samedocument-crossorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/open-samedocument-crossorigin-sameorigindomain.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/open-samedocument-sameorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/submit-samedocument-crossorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/submit-samedocument-crossorigin-sameorigindomain.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/submit-samedocument-sameorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-after-dispatch-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-canceled-event-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-handler-null-or-undefined-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-anchor-fragment-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-anchor-userInitiated-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-form-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-form-get-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-form-userInitiated-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-back-after-fragment-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-back-noop-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-go-0-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-iframe-location-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-location-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-meta-refresh-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-navigation-back-same-document-in-iframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-navigation-navigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-svg-anchor-fragment-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-window-open-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-window-open-self-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/same-url-replace-cross-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/same-url-replace-same-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-history-push-not-loaded-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-same-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/scroll-behavior/manual-scroll-after-resolve-expected.txt:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::determineNavigationType):
(WebCore::FrameLoader::loadInSameDocument):
* Source/WebCore/page/NavigateEvent.cpp:
(WebCore::NavigateEvent::NavigateEvent):
(WebCore::NavigateEvent::create):
(WebCore::NavigateEvent::sharedChecks):
(WebCore::NavigateEvent::intercept):
(WebCore::NavigateEvent::scroll):
(WebCore::NavigateEvent::finish):
* Source/WebCore/page/NavigateEvent.h:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::currentEntry const):
(WebCore::Navigation::updateCurrentEntry):
(WebCore::Navigation::hasEntriesAndEventsDisabled const):
(WebCore::Navigation::updateForNavigation):
(WebCore::documentCanHaveURLRewritten):
(WebCore::Navigation::innerDispatchNavigateEvent):
(WebCore::Navigation::dispatchTraversalNavigateEvent):
(WebCore::Navigation::dispatchPushReplaceReloadNavigateEvent):
(WebCore::Navigation::dispatchDownloadNavigateEvent):
(WebCore::determineNavigationType): Deleted.
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/NavigationTransition.h:

Canonical link: <a href="https://commits.webkit.org/276522@main">https://commits.webkit.org/276522@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/64e8f38875466cfbc5419c50472b28bdb7875a37

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44890 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23989 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47380 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47544 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40895 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47193 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28045 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21386 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36852 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45468 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21036 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/38646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17939 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/18447 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/39790 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2937 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41115 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40083 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49215 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/19858 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16398 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/43846 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/21177 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/42608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9992 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/21515 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20851 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/page/NavigationDestination.cpp:
```diff
@@ -32,4 +32,19 @@ namespace WebCore {
+}
+
+bool NavigationDestination::sameDocument() const
+{
```
This may need a FIXME.

rwlbuis commented on Source/WebCore/page/NavigateEvent.h:
```diff
@@ -81,22 +89,30 @@ class NavigateEvent final : public Event {
     bool m_userInitiated { false };
     bool m_hashChange { false };
     bool m_hasUAVisualTransition { false };
+    InterceptionState m_interceptionState { InterceptionState::None };
```
If InterceptionState is not specced, we could remove InterceptionState::None and instead make this a std::optional (to represent None).

rwlbuis commented on Source/WebCore/page/NavigateEvent.cpp:
```diff
@@ -52,12 +53,63 @@ Ref<NavigateEvent> NavigateEvent::create(const AtomString& type, const NavigateE
+ExceptionOr<void> NavigateEvent::intercept(NavigationInterceptOptions&& options)
+{
+    auto checkResult = sharedChecks();
+    if (checkResult.hasException())
```
In modern C++ these 2 lines can be combined to one if IIRC.

TingPing replied:
There are a few ways to format this but I'm not sure what you intended. `checkResult` isn't truthy if an exception is set so `if (auto checkResult = sharedChecks())` doesn't work unfortunately.

rwlbuis replied:
That is one of the things I like about modern C++ :) For example:
https://skebanga.github.io/if-with-initializer/

rwlbuis commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1561,6 +1561,13 @@ void FrameLoader::loadURL(FrameLoadRequest&& frameLoadRequest, const String& ref
         policyChecker().setLoadType(newLoadType);
         RELEASE_ASSERT(!isBackForwardLoadType(newLoadType) || history().provisionalItem());
+
+        RefPtr domWindow = frame->document()->domWindow();
```
We may be able to use the protected variant of the domWindow accessor, if there is one.

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1273,6 +1273,12 @@ void FrameLoader::loadInSameDocument(URL url, RefPtr<SerializedScriptValue> stat
 
     checkedHistory()->updateForSameDocumentNavigation();
 
+    RefPtr domWindow = m_frame->document()->protectedWindow();
```
We already have document.

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1273,6 +1273,12 @@ void FrameLoader::loadInSameDocument(URL url, RefPtr<SerializedScriptValue> stat
 
+    RefPtr domWindow = m_frame->document()->protectedWindow();
+    if (domWindow && m_frame->document()->settings().navigationAPIEnabled()) {
+        domWindow->protectedNavigation()->dispatchNavigationEvent(url, oldURL, documentLoader()->triggeringAction().type());
```
Will this affect non-users of the new API when the new API is turned on?

TingPing replied:
Any state it touches is contained to the Navigation object. So it shouldn't outside of bugs internal to Navigation, but the test coverage and asserts should be decent.

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -375,4 +378,31 @@ void Navigation::updateForNavigation(Ref<HistoryItem>&& item, FrameLoadType fram
         disposedEntry->dispatchEvent(Event::create(eventNames().disposeEvent, { }));
 }
 
+bool Navigation::dispatchNavigationEvent(const URL& newURL, const URL& oldURL, FrameLoadType loadType)
```
it would be good to have a reference link here. Also should it be dispatchNavigateEvent?

rwlbuis commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1275,8 +1275,11 @@ void FrameLoader::loadInSameDocument(URL url, RefPtr<SerializedScriptValue> stat
+            if (RefPtr currentItem = checkedHistory()->currentItem()) {
+                domWindow->protectedNavigation()->dispatchNavigationEvent(url, oldURL, policyChecker().loadType());
+                // FIXME: Properly handle result of navigate event.
                 domWindow->protectedNavigation()->updateForNavigation(currentItem.releaseNonNull(), policyChecker().loadType());
```
It looks like these two operations are quite far apart in the specification:
https://html.spec.whatwg.org/multipage/browsing-the-web.html#apply-the-history-step

Did you consider a different place for sending the navigate Event? If we can't do it right now but want to later then maybe we can add a FIXME.

rwlbuis reviewed (commented):
Looks mostly good, left some comments.

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -379,4 +409,186 @@ void Navigation::updateForNavigation(Ref<HistoryItem>&& item, FrameLoadType fram
+    if (!documentOrigin.isSameSiteAs(targetOrigin))
+        return false;
+
+    if (targetURL.protocolIs("https"_s))
```
I think this should check "http" too.

rwlbuis reviewed (commented):
I think the PR is too big, making review difficult. How about:
1. refactor PR, with NavigationTransition + NavigationDestination changes + scriptExecutionContext()  cleanup.
2. just a PR with event dispatch algorithms
3. a PR that calls the event dispatch algorithms

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -379,4 +365,159 @@ void Navigation::updateForNavigation(Ref<HistoryItem>&& item, FrameLoadType fram
+
+    return equalIgnoringFragmentIdentifier(documentURL, targetURL);
+}
+
```
extra space.

achristensen07 commented on Source/WebCore/page/NavigateEvent.h:
```diff
@@ -81,22 +90,38 @@ class NavigateEvent final : public Event {
+    std::optional<NavigationFocusReset> m_focusReset;
+    std::optional<NavigationScrollBehavior> m_scrollBehavior;
+    RefPtr<AbortController> m_abortController;
+
```
extra space

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -379,4 +365,159 @@ void Navigation::updateForNavigation(Ref<HistoryItem>&& item, FrameLoadType fram
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#fire-a-download-request-navigate-event
+bool Navigation::dispatchDownloadNavigateEvent(const URL&, const String& downloadFilename)
+{
+    // FIXME
```
When to call this is tricky, since it seems to be handled mostly in FrameLoaderClient. Maybe an option is to send the event in PolicyChecker::checkNavigationPolicy before calling startDownload.

TingPing replied:
Yeah, the policychecker was one place I was looking.

All of these dispatches seem tricky to find the ideal spot, the general web navigation specs don't seem super aligned with WebKit's structure to me.

rwlbuis replied:
> Yeah, the policychecker was one place I was looking.
> 
> All of these dispatches seem tricky to find the ideal spot, the general navigation specs don't seem super aligned with WebKit's structure to me.

Correct, for example the download handling has a quite seperate algorithm, but I would bet that spec was done after the implementation. I would like to work on refactoring that, but it is not crucial for Navigation API.

TingPing committed:
[Navigation] Dispatch NavigateEvent on same-document navigations
https://bugs.webkit.org/show_bug.cgi?id=265401

Reviewed by Alex Christensen.

- Partial implementation intercept/finish/scroll/abort in NavigateEvent.
- Implement a large part of the dispatching of Traversal and Push/Reload/Replace events.
- The FrameLoader dispatches the events in the case of same-document loads.
- Gate NavigateEvent.hasUAVisualTransition behind the setting for it.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/commit-behavior/multiple-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/click-samedocument-crossorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/click-samedocument-crossorigin-sameorigindomain.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/click-samedocument-sameorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/location-samedocument-crossorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/location-samedocument-crossorigin-sameorigindomain.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/location-samedocument-sameorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/open-samedocument-crossorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/open-samedocument-crossorigin-sameorigindomain.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/open-samedocument-sameorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/submit-samedocument-crossorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/submit-samedocument-crossorigin-sameorigindomain.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/submit-samedocument-sameorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-after-dispatch-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-canceled-event-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-handler-null-or-undefined-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-anchor-fragment-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-anchor-userInitiated-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-form-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-form-get-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-form-userInitiated-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-back-after-fragment-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-back-noop-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-go-0-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-iframe-location-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-location-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-meta-refresh-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-navigation-back-same-document-in-iframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-navigation-navigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-svg-anchor-fragment-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-window-open-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-window-open-self-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/same-url-replace-cross-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/same-url-replace-same-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-history-push-not-loaded-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-same-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/scroll-behavior/manual-scroll-after-resolve-expected.txt:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::determineNavigationType):
(WebCore::FrameLoader::loadInSameDocument):
* Source/WebCore/page/NavigateEvent.cpp:
(WebCore::NavigateEvent::NavigateEvent):
(WebCore::NavigateEvent::create):
(WebCore::NavigateEvent::sharedChecks):
(WebCore::NavigateEvent::intercept):
(WebCore::NavigateEvent::scroll):
(WebCore::NavigateEvent::finish):
* Source/WebCore/page/NavigateEvent.h:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::currentEntry const):
(WebCore::Navigation::updateCurrentEntry):
(WebCore::Navigation::hasEntriesAndEventsDisabled const):
(WebCore::Navigation::updateForNavigation):
(WebCore::documentCanHaveURLRewritten):
(WebCore::Navigation::innerDispatchNavigateEvent):
(WebCore::Navigation::dispatchTraversalNavigateEvent):
(WebCore::Navigation::dispatchPushReplaceReloadNavigateEvent):
(WebCore::Navigation::dispatchDownloadNavigateEvent):
(WebCore::determineNavigationType): Deleted.
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/NavigationTransition.h:

Canonical link: https://commits.webkit.org/276522@main
